core/src/main/scala/org/apache/spark/util/ListenerBus.scala

```diff
@@ -17,53 +17,222 @@
+import java.util.concurrent.locks.{Condition, ReentrantLock}
 
 import scala.collection.JavaConverters._
 import scala.reflect.ClassTag
 import scala.util.control.NonFatal
 
+import com.google.common.util.concurrent.ThreadFactoryBuilder
+
 import org.apache.spark.internal.Logging
+import org.apache.spark.scheduler.LiveListenerBus
+
+private class ListenerEventExecutor[L <: AnyRef] (listenerName: String, queueCapacity: Int)
+  extends Logging {
+  private val threadFactory = new ThreadFactoryBuilder().setDaemon(true)
+    .setNameFormat(listenerName + "-event-executor")
+    .build()
+  val listeners = new CopyOnWriteArrayList[L]()
+  /** Holds the events to be processed by this listener. */
+  private val eventQueue = new LinkedBlockingQueue[Runnable](queueCapacity)
+  /**
+   * A single threaded executor service guarantees ordered processing
+   * of the events per listener.
+   */
+  private val executorService: ThreadPoolExecutor =
+    new ThreadPoolExecutor(1, 1, 0L, TimeUnit.MILLISECONDS, eventQueue, threadFactory)
+  /** A counter for dropped events. It will be reset every time we log it. */
+  private val droppedEventsCounter = new AtomicLong(0L)
+  /** When `droppedEventsCounter` was logged last time in milliseconds. */
+  private var lastReportTimestamp = 0L
+  /** Indicates if we are in the middle of processing some event */
+  private val processingEvent = new AtomicBoolean(false)
+  /**
+   * Indicates if the event executor is started. The executor thread will be
+   * blocked on the condition variable until the event executor is started to
+   * guarantee that we do not process any event before starting the event executor.
+   */
+  private val isStarted = new AtomicBoolean(false)
+  private val lock = new ReentrantLock()
+  /** Condition variable which is signaled once the event executor is started */
+  private val startCondition: Condition = lock.newCondition
+
+  def addListener(listener: L): Unit = {
+    listeners.add(listener)
+  }
+
+  def removeListener(listener: L): Unit = {
+    listeners.remove(listener)
+  }
+
+  def start(): Unit = {
+    isStarted.set(true)
+    lock.lock()
+    try {
+      startCondition.signalAll()
+    } finally {
+      lock.unlock()
+    }
+  }
+
+  private[this] def waitForStart(): Unit = {
+    lock.lock()
+    try {
+      while (!isStarted.get()) {
+          startCondition.await()
+        }
+    } finally {
+      lock.unlock()
+    }
+  }
+
+  def submit(task: Runnable): Unit = {
+    try {
+      executorService.submit(new Runnable {
+        override def run(): Unit = LiveListenerBus.withinListenerThread.withValue(true) {
+          waitForStart()
+          processingEvent.set(true)
+          task.run()
+          processingEvent.set(false)
+      }
+      })
+    } catch {
+      case e: RejectedExecutionException =>
+        droppedEventsCounter.incrementAndGet()
+        if (System.currentTimeMillis() - lastReportTimestamp >= 60 * 1000) {
```

Comment:
When measuring elapsed time it's generally better to use `System.nanoTime` since it's monotonic.